### ISSUE TITLE
Deno Compatibility

### ISSUE THREAD

**Issue by 7flash** [lgrammel/modelfusion#188]

**Comment by 7flash** [comment]
It works "deno run main.ts"

main.ts
```
import { generateText, openai } from "npm:modelfusion";
```

But when I am trying to build with deno-esbuild, it seems to not support npm specifiers and therefore would be better if I could import modelfusion package directly from deno.land

```
 const { generateText } = await import("https://deno.land/x/modelfusion/mod.ts");
```

**Comment by 7flash** [comment]
As a temporary solution, following works:

```
    const { generateText } = await import('https://dev.jspm.io/modelfusion');  
```

**Comment by lgrammel** [comment]
@7flash thanks for reporting the issue! 

I'm not too familiar with Deno. From your report, my sense is that ideally modelfusion should work with deno-esbuild. Can you tell me the error that you encountered w/ deno-esbuild?

regarding deno.land, is that a separate pkg registry i would need to upload to?

**Comment by 7flash** [comment]
The issue https://github.com/evanw/esbuild/issues/3506 seemed to be resolved by using jspm and though it works for simple case it fails when trying to register tools because of another issue with zod https://github.com/colinhacks/zod/issues/2991

> @7flash thanks for reporting the issue!
> 
> I'm not too familiar with Deno. From your report, my sense is that ideally modelfusion should work with deno-esbuild. Can you tell me the error that you encountered w/ deno-esbuild?
> 
> regarding deno.land, is that a separate pkg registry i would need to upload to?

**Comment by lgrammel** [comment]
https://github.com/colinhacks/zod/issues/2991 is not a Zod error. ModelFusion contains something called `ZodSchema` as well (I want to change the name to disambiguate, but haven't come up with a better name yet).

**Comment by 7flash** [comment]
Is there are a way to provide directy my json schema instead of zod? Because I am building kind of visual editor based on  modelfusion where user ends up describing his tools in json format anyway

**Comment by lgrammel** [comment]
@7flash yes, you can use `new UncheckedSchema` (see `https://github.com/lgrammel/modelfusion/blob/main/examples/basic/src/util/schema/generate-structure-unchecked-json-schema-example.ts` for an example).

This will not do any validation on the structure of the result tho, just cast it (which could be wrong at runtime).

You can also implement your own schema and json schema acccessor, see `https://github.com/lgrammel/modelfusion/blob/main/src/core/schema/Schema.ts` and `https://github.com/lgrammel/modelfusion/blob/main/src/core/schema/JsonSchemaProducer.ts` for the relevant interfaces (they both need to be implemented in a single object / class). That way, you can use other JSON validations like Validbot.

Please note that ModelFusion uses Zod internally for validations that are not related to user-defined schemas.

**Comment by 7flash** [comment]
@lgrammel that's amazing, and here just recorded a demo of visual agent editor I am building with modelfusion: https://youtu.be/V-rI_ughijw

**Comment by lgrammel** [comment]
> @lgrammel that's amazing, and here just recorded a demo of visual agent editor I am building with modelfusion: https://youtu.be/V-rI_ughijw

Pretty cool! Do you have an open source project or a website? I have started a showcase section on the ModelFusion website, when the project is mature and available, this might be a good fit.

**Comment by 7flash** [comment]
> > @lgrammel that's amazing, and here just recorded a demo of visual agent editor I am building with modelfusion: https://youtu.be/V-rI_ughijw
> 
> Pretty cool! Do you have an open source project or a website? I have started a showcase section on the ModelFusion website, when the project is mature and available, this might be a good fit.

Yes it's here https://github.com/galaxydo/galaxy-polkadot

**Comment by lgrammel** [comment]
> > > @lgrammel that's amazing, and here just recorded a demo of visual agent editor I am building with modelfusion: https://youtu.be/V-rI_ughijw
> > 
> > 
> > Pretty cool! Do you have an open source project or a website? I have started a showcase section on the ModelFusion website, when the project is mature and available, this might be a good fit.
> 
> Yes it's here https://github.com/galaxydo/galaxy-polkadot

Awesome - feel free to send a pull request. The "story teller" example in the showcase shows what needs to be added (section & separate page)
